### ISSUE TITLE
remove deprecation warnings from Boost 1.60 in parser

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/stan#1814]
#### Summary:

Boost 1.60 causes g++ to spit out a gazillion warnings for deprecation.

#### Description:

```
> ./runTests.py -j4 src/test/unit/lang/parser/assignment_test.cpp 
```

causes

```
In file included from lib/stan_math/lib/boost_1.60.0/boost/iostreams/detail/is_dereferenceable.hpp:12:0,
                 from lib/stan_math/lib/boost_1.60.0/boost/iostreams/detail/resolve.hpp:26,
                 from lib/stan_math/lib/boost_1.60.0/boost/iostreams/detail/push.hpp:24,
                 from lib/stan_math/lib/boost_1.60.0/boost/iostreams/detail/streambuf/indirect_streambuf.hpp:31,
                 from lib/stan_math/lib/boost_1.60.0/boost/iostreams/stream_buffer.hpp:22,
                 from lib/stan_math/lib/boost_1.60.0/boost/iostreams/stream.hpp:21,
                 from lib/stan_math/lib/boost_1.60.0/boost/spirit/home/qi/stream/detail/iterator_source.hpp:14,
                 from lib/stan_math/lib/boost_1.60.0/boost/spirit/home/qi/stream/stream.hpp:16,
                 from lib/stan_math/lib/boost_1.60.0/boost/spirit/home/qi/stream.hpp:15,
                 from lib/stan_math/lib/boost_1.60.0/boost/spirit/home/qi.hpp:30,
                 from lib/stan_math/lib/boost_1.60.0/boost/spirit/include/qi.hpp:16,
                 from src/stan/lang/grammars/bare_type_grammar_def.hpp:4,
                 from src/stan/lang/grammars/bare_type_grammar_inst.cpp:1:
lib/stan_math/lib/boost_1.60.0/boost/type_traits/detail/bool_trait_def.hpp:18:79: note: #pragma message: NOTE: Use of this header (bool_trait_def.hpp) is deprecated
```

and about a bajillion more, all of which seem to be due to including these through 

* `boost/type_traits/detail/bool_trait_def.hpp`
* `/boost/type_traits/detail/template_arity_spec.hpp`

which are all included through

* `boost/spirit/include/qi.hpp`

#### Reproducible Steps:

Get a g++ compiler and compile the current `develop` branch.

#### Current Output:

See above.

#### Expected Output:

No warnings.

**Comment by bob-carpenter** [comment]
Not sure this is going to be fixable.  They're patching for Boost 1.61, apparently:

http://permalink.gmane.org/gmane.comp.lib.boost.devel/264164

**Comment by bob-carpenter** [comment]
Closing because this isn't something we can fix and they just released Boost 1.61.
